cheevos/cheevos.c

```diff
@@ -1388,27 +1359,92 @@ static int cheevos_test_cond_set(const cheevos_condset_t *condset,
       cheevos_locals.add_buffer = 0;
       cheevos_locals.add_hits   = 0;
 
-      /* Sequential or non-sequential? */
-      set_valid &= cond_valid;
+      if (cond->type == CHEEVOS_COND_TYPE_PAUSE_IF)
+      {
+         /* as soon as we find a PauseIf that evaluates to true,
+          * stop processing the rest of the group. */
+         if (cond_valid)
+            return 1;
+
+         /* if we make it to the end of the function, make sure we are
+          * indicating nothing matched. if we do find a later PauseIf match,
+          * it'll automatically return true via the previous condition. */
+         set_valid = 0;
+
+         if (cond->req_hits == 0)
+         {
+            /* PauseIf didn't evaluate true, and doesn't have a HitCount,
+             * reset the HitCount to indicate the condition didn't match. */
+            if (cond->curr_hits != 0)
+            {
+               cond->curr_hits = 0;
+               *dirty_conds = 1;
+            }
+         }
+         else
+         {
+            /* PauseIf has a HitCount that hasn't been met, ignore it for now. */
+         }
+      }
+      else if (cond->type == CHEEVOS_COND_TYPE_RESET_IF)
+      {
+         if (cond_valid)
+         {
+            *reset_conds = 1; /* Resets all hits found so far */
+            set_valid    = 0; /* Cannot be valid if we've hit a reset condition. */
+         }
+      }
+      else /* Sequential or non-sequential? */
+         set_valid &= cond_valid;
    }
 
-   /* Now, ONLY read reset conditions! */
-   for (cond = condset->conds; cond < end; cond++)
-   {
-      if (cond->type != CHEEVOS_COND_TYPE_RESET_IF)
-         continue;
+   return set_valid;
+}
 
-      cond_valid = cheevos_test_condition(cond);
+static int cheevos_test_pause_cond_set(const cheevos_condset_t *condset,
+      int *dirty_conds, int *reset_conds)
+{
+   if (!condset)
+      return 1; /* important: empty group must evaluate true */
 
-      if (cond_valid)
+   /* identify any Pause conditions and their dependent AddSource/AddHits. */
+   int pause_conds[condset->count];
+   int in_pause              = 0;
+   int has_pause             = 0;
+   cheevos_cond_t *cond      = NULL;
+   const cheevos_cond_t *end = condset->conds + condset->count;
+   int i                     = 0;
+   for (cond = condset->conds; cond < end; cond++)
```

Comment:
AddSource/SubSource/AddHits appear before the condition they modify. You have to iterate over this loop in reverse to set the `pause_conds` correctly for them.

Reply:
done as you suggested:
https://github.com/meleu/RetroArch/blob/22abc6cbb2e454b218b2473dfa2b46d8687e7494/cheevos/cheevos.c#L1412